src/backend/distributed/utils/attribute.c

```diff
@@ -175,60 +182,31 @@ AttributeQueryIfAnnotated(const char *query_string, CmdType commandType)
 
 	if (strncmp(ATTRIBUTE_PREFIX, query_string, strlen(ATTRIBUTE_PREFIX)) == 0)
 	{
-		/* TODO create a function to safely parse the tenant identifier from the query comment */
-		/* query is attributed to a tenant */
-		char *tenantId = (char *) query_string + strlen(ATTRIBUTE_PREFIX);
-		char *tenantEnd = tenantId;
-		while (true && tenantEnd[0] != '\0')
+		char *annotation = ExtractTopComment(query_string);
+		if (annotation != NULL)
 		{
-			if (tenantEnd[0] == ' ' && tenantEnd[1] == '*' && tenantEnd[2] == '/')
+			Datum jsonbDatum = DirectFunctionCall1(jsonb_in, PointerGetDatum(annotation));
+
+			text *tenantIdTextP = ExtractFieldTextP(jsonbDatum, "tId");
+			if (tenantIdTextP != NULL)
 			{
-				break;
+				char *tenantId = UnescapeCommentChars(text_to_cstring(tenantIdTextP));
```

Comment:
consider using postgres functions (e.g. text_substr) as much as possible

unicode is hard

Reply:
I give it a try to use `text_substr` (see [34e1daf](https://github.com/citusdata/citus/pull/6753/commits/34e1dafb181327ccc5571c60e74f9d17d318fea6)) but it failed because of multibyte chars handling (see [failed ci job](https://app.circleci.com/pipelines/github/citusdata/citus/31182/workflows/78e9f365-05ba-413c-887e-6efb3ee2f097/jobs/1030454)) So I reverted it. We're safe with our own substring implementation.